### PR TITLE
feature(hydra): move to formal python:3.8 image

### PR DESCRIPTION
**Pull request by fruch**
centos7 isn't formally support python3
and were getting python3.6 out of iusrepo
and
https://github.com/iusrepo/announce/issues/18
seem like a good reason to move to formal python image

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

File: docker/env/Dockerfile [hunk 1]
```diff
@@ -1,15 +1,28 @@
-FROM centos:7
+FROM python:3.8
 
-ADD install-prereqs.sh install-prereqs.sh
-RUN ./install-prereqs.sh docker
-RUN yum update -y \
-    && yum install -y https://centos7.iuscommunity.org/ius-release.rpm \
-    && yum install -y python36u python36u-libs python36u-devel python36u-pip
+#ADD install-prereqs.sh install-prereqs.sh
```
Comment (fruch) [inline docker/env/Dockerfile:3]: need to remove the comments

File: docker/env/Dockerfile [hunk 2]
```diff
@@ -1,15 +1,28 @@
...
 RUN pip3 install -r requirements-python.txt
 
+# update locales
+RUN apt-get update \
+    && DEBIAN_FRONTEND=noninteractive apt-get install -y locales \
+    && sed -i -e 's/# en_US.UTF-8 UTF-8/en_US.UTF-8 UTF-8/' /etc/locale.gen \
+    && dpkg-reconfigure --frontend=noninteractive locales \
+    && update-locale LANG=en_US.UTF-8
+
 ENV LANG en_US.UTF-8
```
Comment (fruch) [inline docker/env/Dockerfile:26]: check if those are needed, since `update-locale LANG=en_US.UTF-8` might handle those automatically already

File: docker/env/version [hunk 3]
```diff
@@ -1 +1 @@
-0.66
+0.67-python3.8
```
Comment (fruch) [inline docker/env/version:1]: drop python3.8 out of the name,
this is now only for testing, so I can delete those images more clearly

File: unit_tests/test_seed_selector.py [hunk 4]
```diff
@@ -114,7 +114,7 @@ def test_reflector_node_not_exists(self):
         self.setup_cluster(nodes_number=1)
         self.cluster.set_test_params(seeds_selector='reflector', seeds_num=1, db_type='scylla')
         sdcm.cluster.SCYLLA_YAML_PATH = os.path.join(os.path.dirname(__file__), 'test_data', 'scylla.yaml')
-        self.assertRaisesRegex(expected_exception=RetryError,
-                               expected_regex='Waiting for seed is selected by reflector',
-                               callable_obj=self.cluster.set_seeds,
+        self.assertRaisesRegex(RetryError,
```
Comment (bentsi) [inline unit_tests/test_seed_selector.py:117]: why did you remove argument names?

Comment (fruch) [reply unit_tests/test_seed_selector.py:117]: `callable_obj` parameter was removed in 3.8
and can't do unnamed argument after named ones, so needed to take out the other too

**Comment by fruch** [comment]
Status:
docker provision test passing locally
all provision tests in jenkins seem to be failing,
AWS:
```
[2020-05-21T20:01:44.358Z] < t:2020-05-21 20:01:43,847 f:log.py          l:20   c:sdcm.utils.log       p:ERROR >   File "/sct/sdcm/utils/common.py", line 1073, in get_my_ip
[2020-05-21T20:01:44.358Z] < t:2020-05-21 20:01:43,847 f:log.py          l:20   c:sdcm.utils.log       p:ERROR >     ip = socket.gethostbyname(hostname)
[2020-05-21T20:01:44.358Z] < t:2020-05-21 20:01:43,847 f:log.py          l:20   c:sdcm.utils.log       p:ERROR > socket.gaierror: [Errno -2] Name or service not known
```
GCE: ssh issue to connect to loader, seem like not related
docker:
import yaml fails when running configen.py, very weird since local with hydra it was running good

**Comment by fruch** [comment]
Status:
docker and AWS provision tests are passing
GCE is failing, doesn't sounds related, trying again to be sure.

**Comment by bentsi** [comment]
remove
- @cached_property
- sdcm/utils/fromisoformat.py

**Comment by bentsi** [comment]
@fruch when running `hydra list-resources --test-id dc1d9454-42b0-4b0a-8346-aa3aa443d548`
Got an error:
```
Docker version 19.03.6, build 369ce74a3c
Image up-to-date
Obtaining QA SSH keys...
QA private key '/home/bentsi/.ssh/scylla-qa-ec2' exists. Nothing to update.
QA private key '/home/bentsi/.ssh/scylla-test' exists. Nothing to update.
QA private key '/home/bentsi/.ssh/support' exists. Nothing to update.
QA SSH keys obtained.
Making sure the ownerships of results directories are of the user
Checking AWS EC2...
Nothing found for selected filters in AWS!
Checking AWS Elastic IPs...
No elastic ips found for selected filters in AWS!
Checking GCE...
Nothing found for selected filters in GCE!
Checking Docker...
Traceback (most recent call last):
  File "./sct.py", line 586, in <module>
    cli()
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "./sct.py", line 228, in list_resources
    list_resources_docker(tags_dict=params, running=get_all_running, group_as_builder=True, verbose=verbose)
  File "/sct/sdcm/utils/common.py", line 540, in list_resources_docker
    docker_clients = tuple(list_clients_docker(builder_name=builder_name, verbose=verbose).items())
  File "/sct/sdcm/utils/common.py", line 494, in list_clients_docker
    SSHAgent.start(verbose=verbose)
  File "/sct/sdcm/utils/ssh_agent.py", line 41, in start
    os.environ.update(json.loads(res.stderr))
  File "/usr/local/lib/python3.8/json/__init__.py", line 357, in loads
    return _default_decoder.decode(s)
  File "/usr/local/lib/python3.8/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/local/lib/python3.8/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

```

**Comment by fruch** [comment]
> @fruch when running `hydra list-resources --test-id dc1d9454-42b0-4b0a-8346-aa3aa443d548`
> Got an error:
> 
> ```
> Docker version 19.03.6, build 369ce74a3c
> Image up-to-date
> Obtaining QA SSH keys...
> QA private key '/home/bentsi/.ssh/scylla-qa-ec2' exists. Nothing to update.
> QA private key '/home/bentsi/.ssh/scylla-test' exists. Nothing to update.
> QA private key '/home/bentsi/.ssh/support' exists. Nothing to update.
> QA SSH keys obtained.
> Making sure the ownerships of results directories are of the user
> Checking AWS EC2...
> Nothing found for selected filters in AWS!
> Checking AWS Elastic IPs...
> No elastic ips found for selected filters in AWS!
> Checking GCE...
> Nothing found for selected filters in GCE!
> Checking Docker...
> Traceback (most recent call last):
>   File "./sct.py", line 586, in <module>
>     cli()
>   File "/usr/local/lib/python3.8/site-packages/click/core.py", line 764, in __call__
>     return self.main(*args, **kwargs)
>   File "/usr/local/lib/python3.8/site-packages/click/core.py", line 717, in main
>     rv = self.invoke(ctx)
>   File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1137, in invoke
>     return _process_result(sub_ctx.command.invoke(sub_ctx))
>   File "/usr/local/lib/python3.8/site-packages/click/core.py", line 956, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>   File "/usr/local/lib/python3.8/site-packages/click/core.py", line 555, in invoke
>     return callback(*args, **kwargs)
>   File "/usr/local/lib/python3.8/site-packages/click/decorators.py", line 17, in new_func
>     return f(get_current_context(), *args, **kwargs)
>   File "./sct.py", line 228, in list_resources
>     list_resources_docker(tags_dict=params, running=get_all_running, group_as_builder=True, verbose=verbose)
>   File "/sct/sdcm/utils/common.py", line 540, in list_resources_docker
>     docker_clients = tuple(list_clients_docker(builder_name=builder_name, verbose=verbose).items())
>   File "/sct/sdcm/utils/common.py", line 494, in list_clients_docker
>     SSHAgent.start(verbose=verbose)
>   File "/sct/sdcm/utils/ssh_agent.py", line 41, in start
>     os.environ.update(json.loads(res.stderr))
>   File "/usr/local/lib/python3.8/json/__init__.py", line 357, in loads
>     return _default_decoder.decode(s)
>   File "/usr/local/lib/python3.8/json/decoder.py", line 337, in decode
>     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
>   File "/usr/local/lib/python3.8/json/decoder.py", line 355, in raw_decode
>     raise JSONDecodeError("Expecting value", s, err.value) from None
> json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
> ```

I've run 
```
hydra list-resources --get-all
```
with this image and didn't seen any issues.

**Comment by fruch** [comment]
**UPDATE**:
the GCE issue was pinpointed to scylla-bench issue
https://github.com/scylladb/scylla-bench/pull/35

merged and released

**Comment by fruch** [comment]
@enaydanov have you seen failures like @bentsi is describing ? the only thing I can think of is that `ssh-agent -s` failed, but we don't have log files from those commands yet to check it out.

**Comment by fruch** [comment]
@bentsi from my POV this one is ready,
your call on when to merge it.

**Comment by bentsi** [comment]
@fruch please add documentation of how to install Python 3.8 on your local env

File: docker/env/Dockerfile [hunk 5]
```diff
@@ -1,15 +1,25 @@
-FROM centos:7
+FROM python:3.8
 
-ADD install-prereqs.sh install-prereqs.sh
-RUN ./install-prereqs.sh docker
-RUN yum update -y \
-    && yum install -y https://repo.ius.io/ius-release-el7.rpm \
-    && yum install -y python36u python36u-libs python36u-devel python36u-pip
+RUN DEBIAN_FRONTEND=noninteractive apt update && \
```
Comment (enaydanov) [inline docker/env/Dockerfile:3]: It's better to change `apt` to `apt-get`. Debian shows warning if you try to run `apt` in scripts.

File: docker/env/Dockerfile [hunk 6]
```diff
@@ -1,15 +1,25 @@
-FROM centos:7
+FROM python:3.8
 
-ADD install-prereqs.sh install-prereqs.sh
-RUN ./install-prereqs.sh docker
```
Comment (enaydanov) [inline docker/env/Dockerfile:4]: It'll be good to update `docker/env/build_n_push.sh` too.

File: docker/env/Dockerfile [hunk 7]
```diff
@@ -1,15 +1,25 @@
...
+RUN DEBIAN_FRONTEND=noninteractive apt update && \
+    apt install -y git openssl iproute2 sudo apt-transport-https \
+    ca-certificates curl gnupg2 software-properties-common rsync
+
+# Install Docker cli only
+RUN curl -fsSL https://download.docker.com/linux/debian/gpg | sudo apt-key add - && \
+    DEBIAN_FRONTEND=noninteractive add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/debian $(lsb_release -cs) stable" && \
+    DEBIAN_FRONTEND=noninteractive apt-get update && apt-get install -y docker-ce-cli  && \
+    sudo groupadd docker && \
+    sudo usermod -aG docker `whoami`
```
Comment (enaydanov) [inline docker/env/Dockerfile:12]: I think you can replace `whoami` with `$USER`

File: docker/env/Dockerfile [hunk 8]
```diff
@@ -1,15 +1,25 @@
...
 RUN pip3 install -r requirements-python.txt
 
-ENV LANG en_US.UTF-8
-ENV LC_ALL en_US.UTF-8
+# update locales
+RUN apt-get update \
+    && DEBIAN_FRONTEND=noninteractive apt-get install -y locales \
+    && sed -i -e 's/# en_US.UTF-8 UTF-8/en_US.UTF-8 UTF-8/' /etc/locale.gen \
+    && dpkg-reconfigure --frontend=noninteractive locales \
+    && update-locale LANG=en_US.UTF-8
```
Comment (enaydanov) [inline docker/env/Dockerfile:23]: Why we need to update locales? It's set to C.UTF-8 and probably we don't need to change it.

Comment (fruch) [reply docker/env/Dockerfile:23]: we had those before, and it was keep supplying warnings,
I've looked for the proper way of setting it, but you are right that this should be the default anyhow.

I'll remove

File: docker/env/Dockerfile (hunk 6 above)
Comment (fruch) [reply docker/env/Dockerfile:4]: to update what exactly ? I'm not sure I understood what do you mean here...

File: docker/env/Dockerfile (hunk 7 above)
Comment (fruch) [reply docker/env/Dockerfile:12]: Tried it, it doesn't work. you need to define this environment variable for it to work.

maybe it's not needed at all..., since in hydra we pass all the group ids into the docker commandline

Comment (enaydanov) [reply docker/env/Dockerfile:12]: Yes, you're right. I was mistakenly ensured that it should work.

File: docker/env/Dockerfile (hunk 6 above)
Comment (enaydanov) [reply docker/env/Dockerfile:4]: You removed `install-prereqs.sh` from Dockerfile, but not from `build_n_push.sh`:  the latter script copies/removes `install-prereqs.sh` to run it in Dockerfile.

Comment (fruch) [reply docker/env/Dockerfile:4]: now removed it

**Comment by fruch** [comment]
@bentsi anything else holding this one ?

**Comment by bentsi** [comment]
@fruch Thursday :D
Let's merge on Sunday

**Comment by fruch** [comment]
@bentsi Friday is a good day for those kind of merges :+1:

**Comment by fruch** [comment]
@bentsi we agreed we'll merge this week, right ?

**Comment by bentsi** [comment]
yes

**Comment by bentsi** [comment]
I will check it with the new Hyra and SCT runners

**Comment by fruch** [comment]
@bentsi a reminder, it's been two week since we agreed to merge this one...
